docs/memorymanagement.md

```diff
@@ -0,0 +1,108 @@
+# C#/WinRT Object Lifetime and Reference Tracking
+
+## Overview
+
+C#/WinRT is a WinRT projection for C# which at a high level generates wrapper C# types to represent
+WinRT types. The lifetime of any of these instantiated C# types is managed by the .NET garbage collector
+as with any C# object. But as a WinRT projection, the lifetime of the WinRT objects it wraps is
+managed by COM reference tracking. The XAML runtime also manages the lifetime of XAML / WinUI objects
+and has its own reference tracking that interacts with .NET and its garbage collector. This document
+serves the purpose of documenting how C#/WinRT interacts with all 3 systems to correctly manage the
+lifetime of projected WinRT objects.
+
+### COM reference tracking
+
+Each WinRT object that we project is based on COM and implements a set of interfaces.
+As per the COM design, every COM interface implements `IUnknown` which has an `AddRef` and `Release`
+function. C#/WinRT calls the `AddRef` function anytime it gets a new reference to a WinRT object
+which C#/WinRT holds onto using an `IObjectReference` instance. It also calls `AddRef` whenever it gives
+out a reference to one of these objects across the ABI as an out parameter. C#/WinRT calls the
+`Release` function whenever any of the `IObjectReference` instances holding onto the WinRT object is
+disposed or finalized by the .NET garbage collector. As long as there are still references to the
+native object, it stays alive even if the projected C# object gets finalized due to there being
+no more references to it from C# code. But if the C# reference was the last reference to it, the
+release will also end up cleaning up the native object.
+
+The above describes what typically happens for any natively implemented WinRT object that C#/WinRT
+projects. There are some differences to this when the object is instead a C# implemented object that is
+projected into WinRT via a COM callable wrapper (CCW). This is done via a C# class implementing a set of
+WinRT interfaces or via a C# class extending (aggregating in COM) an unsealed WinRT type.
+
+In the former, the object is implemented purely in C# and its lifetime is managed by the .NET
+garbage collector. C#/WinRT only comes into play when this C# object is passed across the ABI to a
+WinRT function. When this happens, C#/WinRT creates a CCW for it using the .NET 5 ComWrappers API

```

Comment:
Link to the `ComWrappers` API? https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.comwrappers